libraries/rib-compose/src/main/java/com/badoo/ribs/compose/ComposeRibView.kt

```diff
@@ -2,66 +2,57 @@ package com.badoo.ribs.compose
-import com.badoo.ribs.android.AndroidRibViewHost
 import com.badoo.ribs.core.Node
 import com.badoo.ribs.core.view.RibView
 
+
+/**
+ * RibView used for compose interoperability. It will wrap the view in a ComposeView.
+ *
+ * @param context Android context used to create the required views
+ * @param targetFactory Factory to provide ComposeTarget, required internally to attach compose content
+ * to a ViewGroup. Use a custom one if you need to customize the ViewGroup.
+ */
 abstract class ComposeRibView(
-    override val context: Context
+    override val context: Context,
+    private val targetFactory: (MutableState<ComposeView?>, context: Context) -> ComposeTarget = { state, context ->
+        ComposeTarget(state, context)
+    },
 ) : RibView {
 
     abstract val composable: ComposeView
 
-    /**
-     * Only for compatibility with [com.badoo.ribs.core.view.AndroidRibView] parents.
-     *
-     * Will not be constructed / used if parent is also [ComposeRibView]. In those cases
-     * current [composable] is used directly.
-     */
-    override val androidView: ViewGroup by lazy {
+    override val androidView: ViewGroup by lazy(LazyThreadSafetyMode.NONE) {
         androidx.compose.ui.platform.ComposeView(context).apply {
             setContent(composable)
         }
     }
+    private val targets = mutableMapOf<MutableState<ComposeView?>, ComposeTarget>()
+
+    private fun getTargetForSubtree(subtreeOf: Node<*>): ComposeTarget {
+        val state = getParentStateForSubtree(subtreeOf)
+        return targets.getOrPut(
+            state,
+            { targetFactory(state, context) }
+        )
+    }
 
-    private var lastChildAttached: MutableMap<Any, Node<*>> = mutableMapOf()
-
-    protected open fun getParentViewForSubtree(subtreeOf: Node<*>): MutableState<ComposeView?> =
+    open fun getParentStateForSubtree(subtreeOf: Node<*>): MutableState<ComposeView?> =
         mutableStateOf(null)
 
     override fun attachChild(child: Node<*>, subtreeOf: Node<*>) {
-        val target: MutableState<ComposeView?> = getParentViewForSubtree(subtreeOf)
-        lastChildAttached[target] = child
-
-        when (val childView = child.onCreateView(this)) {
-            is ComposeRibView -> {
-                child.onAttachToView()
-                target.value = childView.composable
-            }
-
-            else -> {
-                val innerContainer = FrameLayout(context)
-                AndroidRibViewHost(innerContainer).attachChild(child)
-                target.value = { AndroidView(factory = { innerContainer }) }
-            }
-        }
+        val target = getTargetForSubtree(subtreeOf)
+        child.onCreateView(this)
+        child.view?.let { target.addView(it.androidView) }
+        child.onAttachToView()
     }
 
+
     override fun detachChild(child: Node<*>, subtreeOf: Node<*>) {
+        val target = getTargetForSubtree(subtreeOf)
+        child.view?.let { target.removeView(it.androidView) }
```

Comment:
```kotlin
val childView = child.onDetachFromView()
target.removeView(childView.androidView)
```

Reply:
Nice idea, I will take a look to the effect of detaching before removing target  👍

Reply:
What about a lambda?     
```
`fun onDetachFromView(removeView: ((view: V)-> Unit)? = null) {
  view?.let {
            removeView?.invoke(it)
        }
...
}

```
Then just call:
`child.onDetachFromView{ target.removeView(it.androidView) }

Reply:
I think lambda looks worse and less obvious, but if you find any issues with the original approach, use it.